python/spark_sklearn/keyed_models.py

```diff
@@ -0,0 +1,407 @@
+    appropriately matched.
+
+    Currently, the class provides a generalization for scikit-learn transformers and predictors
+    (per-key clustering is currently unsupported). Because these scikit-learn estimators
+    all derive from the same base type (yielding the same API), yet have different expectations
+    for what methods should be called and with what arguments, this class enumerates two different
+    types of behavior:
+
+    1. `"transformer"`
+
+        Examples: `sklearn.decomposition.PCA`, `sklearn.cluster.KMeans`
+
+        In this case, the estimator will aggregate the all input features for a given key into a
+        `NxD` data matrix, where `N` is the number of distinct rows for the given key and `D` is the
+        feature space dimensionality; let this matrix be `X`.
+
+        For each such key and data matrix pair, a clone of the parameter estimator is fitted with
+        `estimator.fit(X)`, inducing a mapping between keys and fitted estimators: this produces
+        a fitted transformer :class:`KeyedModel`, whose Spark ML `transform` method generates an
+        output column by applying each key's fitted scikit-learn estimator's own `transform` method.
+
+        The output column type for transformers will always be a :class:`DenseVector`.
+
+    2. `"predictor"`
+
+        Examples: `sklearn.svm.LinearSVC`, `sklearn.linear_model.ElasticNet`
+
+        Here, the estimator will likewise aggregate input features into the data matrix `X`.
+        In addition, the label column will be aggregated in a collated manner, generating
+        a vector `y` for each key. The estimator clone will be fitted with `estimator.fit(X, y)`.
+
+        A predictor :class:`KeyedModel` `transform`s its input dataframe by generating an output
+        column with the output of the estimator's `predict` method.
+
+        The output column type for predictors will be the same as the label column (which
+        must be an :class:`AtomicType` (else a `TypeError` will be thrown at `fit()`-time).
+
+    The input column should be numeric or a vector (else a `TypeError` will be thrown at
+    `fit()`-time). Don't use "estimator" as a column name.
+
+    .. :note: Clustering is *not* supported. For instance, the scikit-learn `KMeans` estimator
+    transforms its data into the cluster-mean-distance space - transform will not generate the
+    column of `KMeans` cluster labels.
+    .. :note: While key-based grouping occurs during training, during the transformation/prediction
+    phase of computation, no such aggregation occurs: the number of rows inputted as test data
+    will be equal to the number of rows outputted.
+    .. :note: `spark.conf.get("spark.sql.retainGroupColumns")` assumed to be `u"true"` for both the
+    keyed estimator and the keyed model.
+    .. :note: Estimators trained, persisted, and loaded across different sklearn versions are not
+    guarnteed to work.
+    """
+
+    paramSpecs = {
```

Comment:
Admittedly weird, but it seemed like the most convenient and appropriate for containing parameter values: I need a collection of the parameter names, a mapping from param names to their documentation, and a mapping from them to their defaults. I can make everything a separate attribute, but it'd make for a lot of attributes. I think that'd be cumbersome to read through, but your call.